components/backup/src/writer.rs

```diff
@@ -42,77 +111,75 @@ impl BackupWriter {
-                        let data_key_default = keys::data_key(&default.0);
-                        self.default.put(&data_key_default, &default.1)?;
-                        self.default_written = true;
+                        self.default.write(&default.0, &default.1)?;
+                        value_in_default = true;
                     }
                     assert!(!write.0.is_empty());
-                    let data_key_write = keys::data_key(&write.0);
-                    self.write.put(&data_key_write, &write.1)?;
-                    self.write_written = true;
+                    self.write.write(&write.0, &write.1)?;
                 }
                 TxnEntry::Prewrite { .. } => {
-                    return Err(Error::Other("prewrite is not supported".into()))
+                    return Err(Error::Other("prewrite is not supported".into()));
                 }
             }
+            if value_in_default {
+                self.default.update_with(e, need_checksum)?;
+            } else {
+                self.write.update_with(e, need_checksum)?;
+            }
         }
         Ok(())
     }
 
     /// Save buffered SST files to the given external storage.
-    pub fn save(mut self, storage: &dyn ExternalStorage) -> Result<Vec<File>> {
-        let name = self.name;
-        let save_and_build_file =
-            |cf, mut contents: &[u8], limiter: Option<Arc<IOLimiter>>| -> Result<File> {
-                BACKUP_RANGE_SIZE_HISTOGRAM_VEC
-                    .with_label_values(&[cf])
-                    .observe(contents.len() as _);
-                let name = format!("{}_{}.sst", name, cf);
-                let checksum = tikv_util::file::calc_crc32_bytes(contents);
-                let mut limit_reader = LimitReader::new(limiter, &mut contents);
-                storage.write(&name, &mut limit_reader)?;
-                let mut file = File::new();
-                file.set_crc32(checksum);
-                file.set_name(name);
-                Ok(file)
-            };
+    pub fn save(self, storage: &dyn ExternalStorage) -> Result<Vec<File>> {
         let start = Instant::now();
         let mut files = Vec::with_capacity(2);
         let mut buf = Vec::new();
-        if self.default_written {
+        let write_written = !self.write.is_empty() || !self.default.is_empty();
+        if !self.default.is_empty() {
             // Save default cf contents.
-            buf.reserve(self.default.file_size() as _);
-            self.default.finish_into(&mut buf)?;
-            let default = save_and_build_file(CF_DEFAULT, &mut buf, self.limiter.clone())?;
+            let default = self
+                .default
+                .save_and_build_file(
+                    &self.name,
+                    CF_DEFAULT,
+                    &mut buf,
+                    self.limiter.clone(),
+                    storage,
+                )
+                .unwrap();
```

Comment:
```suggestion
                ?;
```

Could you return error instead? `unwrap` may cause panic.

Reply:
fixed